_posts/2016-05-13-a-comprehensive-guide-to-interacting-with-imap-using-ruby.markdown

```diff
@@ -0,0 +1,260 @@
+ #<struct Net::IMAP::MailboxList attr=[:Drafts, :Hasnochildren], delim="/", name="[Gmail]/Drafts">,
+ #<struct Net::IMAP::MailboxList attr=[:Hasnochildren, :Important], delim="/", name="[Gmail]/Important">,
+ #<struct Net::IMAP::MailboxList attr=[:Hasnochildren, :Sent], delim="/", name="[Gmail]/Sent Mail">,
+ #<struct Net::IMAP::MailboxList attr=[:Hasnochildren, :Junk], delim="/", name="[Gmail]/Spam">,
+ #<struct Net::IMAP::MailboxList attr=[:Flagged, :Hasnochildren], delim="/", name="[Gmail]/Starred">,
+ #<struct Net::IMAP::MailboxList attr=[:Hasnochildren, :Trash], delim="/", name="[Gmail]/Trash">]
+```
+
+To search your mailboxes by name, using `*` as a wildcard character:
+
+```ruby
+[5] pry(main)> imap.list("", "*Mail")
+=> [#<struct Net::IMAP::MailboxList attr=[:All, :Hasnochildren], delim="/", name="[Gmail]/All Mail">,
+ #<struct Net::IMAP::MailboxList attr=[:Hasnochildren, :Sent], delim="/", name="[Gmail]/Sent Mail">]
+```
+
+## Selecting a mailbox
+
+Most IMAP interactions require having selected a mailbox before doing them.
+There are two ways to do so:
+
+Read-only:
+
+```ruby
+[6] pry(main)> imap.examine("[Gmail]/All Mail")
+=> #<struct Net::IMAP::TaggedResponse
+ name="OK",
+ raw_data="RUBY0014 OK [READ-ONLY] [Gmail]/All Mail selected. (Success)\r\n">
+```
+
+Read and write:
+
+```ruby
+[7] pry(main)> imap.select("[Gmail]/All Mail")
+=> #<struct Net::IMAP::TaggedResponse
+ name="OK",
+ raw_data="RUBY0015 OK [READ-WRITE] [Gmail]/All Mail selected. (Success)\r\n">
+```
+
+## Searching email messages
+
+After selecting an email box, you can search email messages and return either
+the sequence number (`seqno`) for the email, or the unique id (`uid`).
+
+The sequence number indicates the *current position* of the email message in the
+mailbox. So this number will be different after email messages before this email
+get moved to a different mailbox, or if they get deleted:
+
+```ruby
+# Search emails which have a subject that includes the word Jedi and return the sequence number.
+[8] pry(main)> imap.search(["SUBJECT", "Jedi"])
+=> [12]
+
+# From within Gmail, we delete one of our very first emails in the All Mail box and then make the same search we did before.
+
+[9] pry(main)> imap.search(["SUBJECT", "Jedi"])
+=> [11]
+```
+
+The unique id is a unique identifier for the email in a mailbox. So even if the
+email's position moves, the `uid` remains the same:
+
+```ruby
+# Search emails which have a subject that includes the word Jedi and return the unique id.
+[10] pry(main)> imap.uid_search(["SUBJECT", "Jedi"])
+=> [23]
+
+# From within Gmail, we delete one of our very first emails in the All Mail box and then make the same search we did before.
+
+[11] pry(main)> imap.uid_search(["SUBJECT", "Jedi"])
+=> [23]
+```
+
+## Fetching email messages
+
+After we selected a mailbox and we know a message's `seqno` or `uid`, we can
+fetch the contents of the email. If you know the `seqno`:
```

Comment:
@mauro-oto You talk about the contents but later you fetch just the `ENVELOPE` - You might want to rephrase the sentence above ^^